c3s_eqc_automatic_quality_control/api.py

```diff
@@ -0,0 +1,168 @@
+    reduced = {k: v for k, v in request.items() if k in _CATALOG_ALLOWED_KEYS}
+    cads_request = {}
+    for d in request["diagnostics"]:
+        if d not in list_diagnostics():
+            request["diagnostics"].remove(d)
+            logging.warning(
+                f"Skipping diagnostic '{d}' since is not available. "
+                "Run 'eqc diagnostics' to see available diagnostics."
+            )
+
+    # Request to CADS are single variable only
+    for var in request["variables"]:
+        reduced["variable"] = var
+        cads_request.update(
+            {
+                var: download.update_request_date(
+                    reduced,
+                    start=request["start"],
+                    stop=request.get("stop"),
+                    switch_month_day=day,
+                )
+            }
+        )
+    return cads_request
+
+
+def _prepare_run_workdir(
+    request: Dict[Any, Any], target_dir: str
+) -> Tuple[pathlib.Path, str, int]:
+    qar_id = request.pop("qar_id")
+    run_n = request.pop("run_n", 0)
+    run_sub = pathlib.Path(target_dir) / qar_id / f"run_{run_n}"
+    logging.info(f"Processing QAR ID: {qar_id} - RUN n.: {run_n}")
+    try:
+        os.makedirs(run_sub)
+    except FileExistsError:
+        logging.warning(
+            f"Run '{run_n}' for qar '{qar_id}' already exists. "
+            "Results will be overwritten."
+        )
+    return run_sub, qar_id, run_n
+
+
+def run_aqc(
+    request: Dict[Any, Any],
+) -> None:
+    cads_request = process_request(request)
+    chunks = request.get("chunks", {"year": 1, "month": 1})
+
+    for var, req in cads_request.items():
+        logging.info(f"Collecting variable '{var}'")
+        data = download.download_and_transform(
+            collection_id=request["collection_id"], requests=req, chunks=chunks
+        )
+
+        # TODO: SANITIZE ATTRS BEFORE SAVING
+        logging.info(f"Saving metadata for variable '{var}'")
+        with open(f"{var}_metadata.yml", "w", encoding="utf-8") as f:
+            f.write(yaml.dump(data.attrs))
+
+        for d in request.get("diagnostics", []):
+            logging.info(f"Processing diagnostic '{d}' for variable '{var}'")
+            diag_ds = getattr(diagnostics, d)(data)
+
+            res = f"{var}_{d}.png"
+            logging.info(f"Saving diagnostic: '{res}'")
+            fig = plot.line_plot(diag_ds.squeeze(), var=var, title=d)
+            fig.write_image(res)
+    return
+
+
+def run(
+    config_file: str,
+    target_dir: str,
+) -> None:
+    with open(config_file, "r", encoding="utf-8") as f:
+        request = yaml.safe_load(f)
+    original_cwd = os.getcwd()
+
+    # Move into qar subfolder
+    run_sub, qar_id, run_n = _prepare_run_workdir(request, target_dir)
+    os.chdir(run_sub)
+
+    try:
+        run_aqc(request)
+    except Exception:
+        logging.error(traceback.format_exc())
+        logging.error(f"QAR ID: {qar_id} - RUN n.: {run_n} failed ")
```

Comment:
```suggestion
    except Exception:
        logging.exception(f"QAR ID: {qar_id} - RUN n.: {run_n} failed ")
```

Reply:
I think traceback is useful to solve errors but I removed it at this stage of development. I'm gonna add it when implementing the use of logfiles for the dashboard

Reply:
accepted